File: membership_card_verifier/packages/verifier-py/src/membership_card_verifier/stages/stage5.py
import base64
import json
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from typing import Any, Literal, Optional

import httpx

from membership_card_verifier.constants import PRESS_REGISTRY_BODY_ENDPOINT
from membership_card_verifier.types import (
    CardEntry,
    FailedCheck,
    IpfsProvider,
    NonComplianceReport,
    RpcProvider,
    VerificationError,
)


@dataclass
class Stage5Result:
    policy_compliant: bool | None | Literal["skipped"]
    policy_match: Optional[bool]
    press_subsequently_revoked: bool
    non_compliance_reported: bool
    errors: list[VerificationError]


async def verify_stage5(
    card_doc: dict[str, Any],
    card_entry: CardEntry,
    card_address: str,
    card_cid: str,
    raw_card_bytes: bytes,
    rpc: RpcProvider,
    ipfs: IpfsProvider,
    config: Any,
) -> Stage5Result:
    errors: list[VerificationError] = []
    failed_checks: list[FailedCheck] = []

    # Step 1: fetch policy snapshot at the immutable policy_id CID
    policy_doc: dict[str, Any]
    try:
        policy_bytes = await ipfs.fetch(card_doc["policy_id"])
        policy_doc = json.loads(policy_bytes.decode("utf-8"))
    except Exception:
        errors.append(
            VerificationError(
                stage=5,
                code="POLICY_FETCH_FAILED",
                message="Could not fetch policy snapshot",
            )
        )
        return Stage5Result(
            policy_compliant=None,
            policy_match=None,
            press_subsequently_revoked=False,
            non_compliance_reported=False,
            errors=errors,
        )

    # Step 2: evaluate card field values against field_definitions
    field_definitions = policy_doc.get("field_definitions", {})
    for field_name, def_ in field_definitions.items():
        if def_.get("required") and field_name not in card_doc:
            failed_checks.append(
                FailedCheck(
                    check="FIELD_POLICY_VIOLATION",
                    field=field_name,
                    detail=f'Required field "{field_name}" is missing',
                )
            )

    # Step 3: check press authorization on-chain
    press_entry = await rpc.get_press_authorization(
        card_entry.policy_address, card_entry.last_press_address
    )

    press_subsequently_revoked = False
    if not press_entry:
        failed_checks.append(
            FailedCheck(
                check="NO_PRESS_AUTHORIZATION",
                detail=f"No press authorization found for policy={card_entry.policy_address} press={card_entry.last_press_address}",
            )
        )
    elif not press_entry.active:
        press_subsequently_revoked = True

    policy_compliant = len(failed_checks) == 0

    # Step 5: non-compliance reporting
    non_compliance_reported = False
    if not policy_compliant:
        ipfs_card_document = (
            base64.urlsafe_b64encode(raw_card_bytes).decode("ascii").rstrip("=")
        )
        verified_at = (
            datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
        )

        report = NonComplianceReport(
            card_address=card_address,
            press_address=card_entry.last_press_address,
            ipfs_card_document=ipfs_card_document,
            ipfs_cid=card_cid,
            failed_checks=failed_checks,
            verified_at=verified_at,
        )

        endpoint = config.registry_endpoint or PRESS_REGISTRY_BODY_ENDPOINT
        try:
            async with httpx.AsyncClient() as client:
                res = await client.post(
                    f"{endpoint}/non-compliance",
                    json=asdict(report),
                )
            non_compliance_reported = res.is_success
            if not res.is_success:
                errors.append(
                    VerificationError(
                        stage=5,
                        code="NON_COMPLIANCE_REPORT_FAILED",
                        message=f"Registry Body returned HTTP {res.status_code}",
                    )
                )
        except Exception as e:
            errors.append(
                VerificationError(
                    stage=5,
                    code="NON_COMPLIANCE_REPORT_FAILED",
                    message=f"Failed to POST non-compliance report: {e}",
                )
            )

    return Stage5Result(
        policy_compliant=policy_compliant,
        policy_match=None,
        press_subsequently_revoked=press_subsequently_revoked,
        non_compliance_reported=non_compliance_reported,
        errors=errors,
    )

File: membership_card_verifier/packages/verifier/src/stages/stage5.ts
import { PRESS_REGISTRY_BODY_ENDPOINT } from "../constants.js";
import { bytesToBase64Url } from "../crypto.js";
import type {
  RpcProvider,
  IpfsProvider,
  CardDocument,
  CardEntry,
  NonComplianceReport,
  FailedCheck,
  VerificationError,
  VerifierConfig,
} from "../types.js";

export interface Stage5Result {
  policy_compliant: boolean | null | "skipped";
  policy_match: boolean | null;
  press_subsequently_revoked: boolean;
  non_compliance_reported: boolean;
  errors: VerificationError[];
}

interface PolicyDocument {
  field_definitions?: Record<string, { required?: boolean; type?: string; [k: string]: unknown }>;
  [key: string]: unknown;
}

export async function verifyStage5(
  cardDoc: CardDocument,
  cardEntry: CardEntry,
  cardAddress: string,
  cardCid: string,
  rawCardBytes: Uint8Array,
  rpc: RpcProvider,
  ipfs: IpfsProvider,
  config: { registryEndpoint?: string | undefined }
): Promise<Stage5Result> {
  const errors: VerificationError[] = [];
  const failedChecks: FailedCheck[] = [];

  // Step 1: fetch policy snapshot at the immutable policy_id CID
  let policyDoc: PolicyDocument;
  try {
    const policyBytes = await ipfs.fetch(cardDoc.policy_id);
    policyDoc = JSON.parse(new TextDecoder().decode(policyBytes)) as PolicyDocument;
  } catch {
    errors.push({ stage: 5, code: "POLICY_FETCH_FAILED", message: "Could not fetch policy snapshot" });
    return {
      policy_compliant: null,
      policy_match: null,
      press_subsequently_revoked: false,
      non_compliance_reported: false,
      errors,
    };
  }

  // Step 2: evaluate card field values against field_definitions
  if (policyDoc.field_definitions) {
    for (const [fieldName, def] of Object.entries(policyDoc.field_definitions)) {
      if (def["required"] && !(fieldName in cardDoc)) {
        failedChecks.push({
          check: "FIELD_POLICY_VIOLATION",
          field: fieldName,
          detail: `Required field "${fieldName}" is missing`,
        });
      }
    }
  }

  // Step 3: check press authorization on-chain
  const pressEntry = await rpc.getPressAuthorization(
    cardEntry.policy_address,
    cardEntry.last_press_address
  );

  let pressSubsequentlyRevoked = false;
  if (!pressEntry) {
    failedChecks.push({
      check: "NO_PRESS_AUTHORIZATION",
      detail: `No press authorization found for policy=${cardEntry.policy_address} press=${cardEntry.last_press_address}`,
    });
  } else if (!pressEntry.active) {
    // Press was revoked after registration — card is still valid, but flag it
    pressSubsequentlyRevoked = true;
  }

  const policyCompliant = failedChecks.length === 0;

  // Step 5: non-compliance reporting
  let nonComplianceReported = false;
  if (!policyCompliant) {
    const report: NonComplianceReport = {
      card_address: cardAddress,
      press_address: cardEntry.last_press_address,
      ipfs_card_document: bytesToBase64Url(rawCardBytes),
      ipfs_cid: cardCid,
      failed_checks: failedChecks,
      verified_at: new Date().toISOString(),
    };

    const endpoint = config.registryEndpoint ?? PRESS_REGISTRY_BODY_ENDPOINT;
    try {
      const res = await fetch(`${endpoint}/non-compliance`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(report),
      });
      nonComplianceReported = res.ok;
      if (!res.ok) {
        errors.push({
          stage: 5,
          code: "NON_COMPLIANCE_REPORT_FAILED",
          message: `Registry Body returned HTTP ${res.status}`,
        });
      }
    } catch (e) {
      errors.push({
        stage: 5,
        code: "NON_COMPLIANCE_REPORT_FAILED",
        message: `Failed to POST non-compliance report: ${String(e)}`,
      });
    }
  }

  return {
    policy_compliant: policyCompliant,
    policy_match: null, // per-call predicate not supported in VerifierConfig; null = not supplied
    press_subsequently_revoked: pressSubsequentlyRevoked,
    non_compliance_reported: nonComplianceReported,
    errors,
  };
}
